File: actin-common/src/main/java/com/hartwig/actin/clinical/feed/lab/LabExtraction.java
package com.hartwig.actin.clinical.feed.lab;

import com.google.common.annotations.VisibleForTesting;
import com.hartwig.actin.clinical.datamodel.ImmutableLabValue;
import com.hartwig.actin.clinical.datamodel.LabValue;
import com.hartwig.actin.clinical.feed.FeedUtil;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public final class LabExtraction {

    private static final Logger LOGGER = LogManager.getLogger(LabExtraction.class);

    private LabExtraction() {
    }

    @NotNull
    public static LabValue extract(@NotNull LabEntry entry) {
        Limits limits = extractLimits(entry.referenceRangeText());
        double value = entry.valueQuantityValue();
        Boolean isOutsideRef = null;
        if (limits.lower() != null || limits.upper() != null) {
            isOutsideRef = (limits.lower() != null && value < limits.lower()) || (limits.upper() != null && value > limits.upper());
        }

        return ImmutableLabValue.builder()
                .date(entry.issued())
                .code(entry.codeCodeOriginal())
                .name(entry.codeDisplayOriginal())
                .comparator(entry.valueQuantityComparator())
                .value(value)
                .unit(entry.valueQuantityUnit())
                .refLimitLow(limits.lower())
                .refLimitUp(limits.upper())
                .isOutsideRef(isOutsideRef)
                .build();
    }

    @NotNull
    @VisibleForTesting
    static Limits extractLimits(@NotNull String referenceRangeText) {
        Double lower = null;
        Double upper = null;
        if (referenceRangeText.contains("-")) {
            String[] limits = referenceRangeText.split("-");
            lower = FeedUtil.parseDouble(limits[0].trim());
            upper = FeedUtil.parseDouble(limits[1].trim());
        } else if (referenceRangeText.contains(">")) {
            int index = referenceRangeText.indexOf(">");
            lower = FeedUtil.parseDouble(referenceRangeText.substring(index + 1).trim());
        } else if (referenceRangeText.contains("<")) {
            int index = referenceRangeText.indexOf("<");
            upper = FeedUtil.parseDouble(referenceRangeText.substring(index + 1).trim());
        } else if (!referenceRangeText.isEmpty()) {
            LOGGER.warn("Could not parse lab value referenceRangeText '{}'", referenceRangeText);
        }

        return new Limits(lower, upper);
    }

    static class Limits {

        @Nullable
        private final Double lower;
        @Nullable
        private final Double upper;

        Limits(@Nullable final Double lower, @Nullable final Double upper) {
            this.lower = lower;
            this.upper = upper;
        }

        @Nullable
        @VisibleForTesting
        Double lower() {
            return lower;
        }

        @Nullable
        @VisibleForTesting
        public Double upper() {
            return upper;
        }
    }
}

File: clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/emc/lab/LabExtraction.kt
package com.hartwig.actin.clinical.feed.emc.lab

import com.hartwig.actin.clinical.feed.emc.FeedParseFunctions
import com.hartwig.actin.datamodel.clinical.LabValue
import org.apache.logging.log4j.LogManager

object LabExtraction {
    private val LOGGER = LogManager.getLogger(LabExtraction::class.java)

    fun extract(entry: LabEntry): LabValue {
        val limits = extractLimits(entry.referenceRangeText)
        val value = entry.valueQuantityValue
        val isOutsideRef = if (limits.lower != null || limits.upper != null) {
            limits.lower != null && value < limits.lower || limits.upper != null && value > limits.upper
        } else null
        return LabValue(
            date = entry.effectiveDateTime,
            code = entry.codeCodeOriginal,
            name = entry.codeDisplayOriginal,
            comparator = entry.valueQuantityComparator,
            value = value,
            unit = LabUnitResolver.resolve(entry.valueQuantityUnit),
            refLimitLow = limits.lower,
            refLimitUp = limits.upper,
            isOutsideRef = isOutsideRef
        )
    }

    private fun extractLimits(referenceRangeText: String): Limits {
        return when {
            referenceRangeText.contains(">") -> {
                val index = referenceRangeText.indexOf(">")
                val lower = FeedParseFunctions.parseDouble(referenceRangeText.substring(index + 1).trim { it <= ' ' })
                Limits(lower, null)
            }

            referenceRangeText.contains("<") -> {
                val index = referenceRangeText.indexOf("<")
                val upper = FeedParseFunctions.parseDouble(referenceRangeText.substring(index + 1).trim { it <= ' ' })
                Limits(null, upper)
            }

            referenceRangeText.contains("-") -> {
                val separatingHyphenIndex = findSeparatingHyphenIndex(referenceRangeText)
                val lower = FeedParseFunctions.parseDouble(referenceRangeText.substring(0, separatingHyphenIndex))
                val upper = FeedParseFunctions.parseDouble(referenceRangeText.substring(separatingHyphenIndex + 1))
                Limits(lower, upper)
            }

            else -> {
                if (referenceRangeText.isNotEmpty()) {
                    LOGGER.warn("Could not parse lab value referenceRangeText '{}'", referenceRangeText)
                }
                Limits(null, null)
            }
        }
    }

    private fun findSeparatingHyphenIndex(referenceRangeText: String): Int {
        return referenceRangeText.indexOfFirst(Char::isDigit).takeIf { it != -1 }
            ?.let { indexOfFirstDigit ->
                referenceRangeText.indexOf("-", startIndex = indexOfFirstDigit).takeIf { it != -1 }
            }
            ?: throw IllegalArgumentException("Could not determine separating hyphen index from $referenceRangeText")
    }

    data class Limits(val lower: Double?, val upper: Double?)
}